### PR TITLE
Disideris/copyfrom waitforcompletion

### PR DESCRIPTION
**Pull request by disideris**
## Summary of the changes / Why this improves CrateDB

The PR contains changes for having a wait_for_completion boolean option in copyfrom and copyto statements and tries to solve #11791 .

There are only small additions for the wait_for_completion flag in the CopyFromPlannerTest .
The executeOrFail method of CopyFromPlan has now a flag check and I don't know if this should be tested.
testCopyFromFileNoWaitForCompletion is added in CopyIntegrationTest but not sure if it is needed or is a good test.
The new EmptyRowConsumer has no tests. Only RowConsumerToResultReceiver was tested and no other RowConsumers.

## Checklist

 - [x] Added an entry in `CHANGES.txt` for user facing changes
 - [x] Updated documentation & `sql_features` table for user facing changes
 - [x] Touched code is covered by tests
 - [x] [CLA](https://crate.io/community/contribute/cla/) is signed
 - [x] This does not contain breaking changes, or if it does:
    - It is released within a major release
    - It is recorded in ``CHANGES.txt``
    - It was marked as deprecated in an earlier release if possible
    - You've thought about the consequences and other components are adapted
      (E.g. AdminUI)

### CODE REVIEW

File: server/src/main/java/io/crate/planner/statement/CopyFromPlan.java [hunk 1]
```diff
@@ -118,9 +124,15 @@ public void executeOrFail(DependencyCarrier dependencies,
...
-            .execute(consumer, plannerContext.transactionContext());
+        JobLauncher jobLauncher = dependencies.phasesTaskFactory()
+            .create(plannerContext.jobId(), List.of(nodeOpTree));
+
+        if (waitForCompletion) {
+            jobLauncher.execute(consumer, plannerContext.transactionContext());
+        } else {
+            jobLauncher.execute(new EmptyRowConsumer(), plannerContext.transactionContext());
+            consumer.accept(InMemoryBatchIterator.of(Row1.ROW_COUNT_UNKNOWN, SENTINEL), null);
+        }
```
Comment (seut) [inline server/src/main/java/io/crate/planner/statement/CopyFromPlan.java:135]: To make this testable, you could create a method with a `BiFunction` parameter to pass in the jobLauncher and in the test case a testing one to avoid creating the `DependencyCarrier` or `JobLauncher` in a unit test.
Then the unit test could verify that the given consumer is called with `-1`.

File: server/src/test/java/io/crate/integrationtests/CopyIntegrationTest.java [hunk 2]
```diff
@@ -108,6 +110,20 @@ public void testCopyFromFileWithJsonExtension() throws Exception {
...
     }
 
+    @Test
+    public void testCopyFromFileNoWaitForCompletion() throws Exception {
+        execute("create table quotes (id int primary key, " +
+                "quote string index using fulltext) with (number_of_replicas = 0)");
+
+        execute("copy quotes from ? with (format='csv', wait_for_completion=false)",
+                new Object[]{copyFilePath + "test_copy_from_csv.ext"});
+        assertThat(response.rowCount(), either(is(0L)).or(is(-1L)));
```
Comment (seut) [inline server/src/test/java/io/crate/integrationtests/CopyIntegrationTest.java:120]: How can this be `0`? 
Anyhow, I think we can create a unit test instead, see my previous comment.

**Comment by seut** [comment]
Looks good in general, I've made some suggestions related to tests.

File: server/src/test/java/io/crate/integrationtests/CopyIntegrationTest.java (hunk 2 above)
Comment (disideris) [reply server/src/test/java/io/crate/integrationtests/CopyIntegrationTest.java:120]: Most of the times is 0! But I ll remove it completely!

Comment (seut) [reply server/src/test/java/io/crate/integrationtests/CopyIntegrationTest.java:120]: If it's `0` there must be a bug as your code always return a `-1` row.

Comment (disideris) [reply server/src/test/java/io/crate/integrationtests/CopyIntegrationTest.java:120]: Yes that was odd. Most times is 0 but some times is -1. How can this happen?

File: server/src/main/java/io/crate/planner/statement/CopyFromPlan.java [hunk 3]
```diff
@@ -118,9 +124,15 @@ public void executeOrFail(DependencyCarrier dependencies,
...
-            .create(plannerContext.jobId(), List.of(nodeOpTree))
-            .execute(consumer, plannerContext.transactionContext());
+        JobLauncher jobLauncher = dependencies.phasesTaskFactory()
+            .create(plannerContext.jobId(), List.of(nodeOpTree));
+
+        if (waitForCompletion) {
+            jobLauncher.execute(consumer, plannerContext.transactionContext());
+        } else {
+            jobLauncher.execute(new EmptyRowConsumer(), plannerContext.transactionContext());
+            consumer.accept(InMemoryBatchIterator.of(Row1.ROW_COUNT_UNKNOWN, SENTINEL), null);
```
Comment (BaurzhanSakhariev) [inline server/src/main/java/io/crate/planner/statement/CopyFromPlan.java:134]: `jobLauncher.execute` might call passed consumer's `accept` in case of exception, see line `consumer.accept(null, throwable)`. @seut, I'm not sure but is it fine in that case, that we are accepting 2 consumers one after one?

File: server/src/main/java/io/crate/planner/statement/CopyFromPlan.java (hunk 1 above)
Comment (disideris) [reply server/src/main/java/io/crate/planner/statement/CopyFromPlan.java:135]: Sorry didn't understand it. Could you rephrase?

Comment (disideris) [reply server/src/main/java/io/crate/planner/statement/CopyFromPlan.java]: Made some changes according to your review.

File: server/src/test/java/io/crate/integrationtests/CopyIntegrationTest.java (hunk 2 above)
Comment (disideris) [reply server/src/test/java/io/crate/integrationtests/CopyIntegrationTest.java]: I removed the integration test but the odd behaviour persisted. Some times was -1 some times 0.

File: server/src/main/java/io/crate/analyze/AnalyzedCopyFrom.java [hunk 4]
```diff
@@ -70,6 +71,14 @@ public Symbol uri() {
         return uri;
     }
 
+    public boolean waitForCompletion() {
+        return waitForCompletion;
+    }
+
+    public void setWaitForCompletion(boolean waitForCompletion) {
```
Comment (mfussenegger) [inline server/src/main/java/io/crate/analyze/AnalyzedCopyFrom.java:78]: A analysed statement should be immutable

File: server/src/main/java/io/crate/analyze/AnalyzedCopyTo.java [hunk 5]
```diff
@@ -79,6 +80,14 @@ public Symbol whereClause() {
         return whereClause;
     }
 
+    public boolean waitForCompletion() {
+        return waitForCompletion;
+    }
+
+    public void setWaitForCompletion(boolean waitForCompletion) {
```
Comment (mfussenegger) [inline server/src/main/java/io/crate/analyze/AnalyzedCopyTo.java:87]: Same here

File: server/src/main/java/io/crate/analyze/AnalyzedCopyFrom.java (hunk 4 above)
Comment (disideris) [reply server/src/main/java/io/crate/analyze/AnalyzedCopyFrom.java:78]: @mfussenegger Ok the set method will be removed.
I ended up having this set method because i was trying to find an optimal way to get the wait_for_completion setting once, after bind method evaluates symbols. Since bind method is static what do you think is the best practice would be to have the wait_for_completion value inside executeOrFail method of copy plans?

File: docs/appendices/release-notes/unreleased.rst [hunk 6]
```diff
@@ -63,6 +63,10 @@ Changes
 - Write blocks added due to low disk space are now automatically removed if a
   node again drops below the high watermark.
 
+- Added a ``WITH`` clause parameter, ``validation`` to ``COPY FROM`` which
```
Comment (BaurzhanSakhariev) [inline docs/appendices/release-notes/unreleased.rst:66]: Probably rebase issues? Was added in https://github.com/crate/crate/pull/12219/files

File: docs/sql/statements/copy-from.rst [hunk 7]
```diff
@@ -558,6 +568,26 @@ An example of using input file with no header
     COPY OK, 3 rows affected (... sec)
 
 
+.. _sql-copy-from-validation:
+
+``validation``
+''''''''''''''
```
Comment (BaurzhanSakhariev) [inline docs/sql/statements/copy-from.rst:574]: same here

File: docs/appendices/release-notes/unreleased.rst (hunk 6 above)
Comment (BaurzhanSakhariev) [reply docs/appendices/release-notes/unreleased.rst:66]: This entry is already on master (see PR above) and should not appear in changes after successful rebase.

File: server/src/test/java/io/crate/integrationtests/CopyIntegrationTest.java (hunk 2 above)
Comment (BaurzhanSakhariev) [reply server/src/test/java/io/crate/integrationtests/CopyIntegrationTest.java]: Is this issue resolved? In the latest revision `CopyIntegrationTest` changes are removed (except imports) - I wonder what happens if we run on the current revision? If this test is not needed (??) then please revert resudual changes in `CopyIntegrationTest`

Comment (disideris) [reply server/src/test/java/io/crate/integrationtests/CopyIntegrationTest.java]: @BaurzhanSakhariev The issue existed as long as the test existed. @seut suggested above to create a unit test instead so i removed it. Unit test works fine and never had an issue with manual tests. 
I will remove the unused imports :)

File: libs/dex/src/testFixtures/java/io/crate/testing/TestingRowConsumer.java [hunk 8]
```diff
@@ -60,8 +60,12 @@ public void accept(BatchIterator<Row> it, Throwable failure) {
     }
 
     public List<Object[]> getResult() throws Exception {
+        return getResultWithTimeout(10);
+    }
+
+    public List<Object[]> getResultWithTimeout(int timeout) throws Exception {
```
Comment (seut) [inline libs/dex/src/testFixtures/java/io/crate/testing/TestingRowConsumer.java:66]: The additional argument already documents the difference to `getResult()`
```suggestion
    public List<Object[]> getResult(int timeout) throws Exception {
```

File: server/src/main/java/io/crate/planner/statement/CopyFromPlan.java [hunk 9]
```diff
@@ -345,6 +347,15 @@ private static void rewriteToCollectToUsePartitionValues(List<Reference> partiti
         }
     }
 
+    private boolean waitForCompletion(ExecutionPlan plan) {
+        try {
+            return ((FileUriCollectPhase) ((Collect) plan).collectPhase()).withClauseOptions()
+                    .getAsBoolean("wait_for_completion", true);
+        } catch (Exception e) {
+            return true;
+        }
```
Comment (seut) [inline server/src/main/java/io/crate/planner/statement/CopyFromPlan.java:356]: What exception can occur here? 
If it's about the a possible cast error, I suggest to protect this with an assertion instead and inline the code in the caller.

File: server/src/main/java/io/crate/planner/statement/CopyPlan.java [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+import io.crate.data.InMemoryBatchIterator;
+import io.crate.data.Row;
+import io.crate.data.Row1;
+import io.crate.data.RowConsumer;
+import io.crate.metadata.TransactionContext;
+
+
+public abstract class CopyPlan {
+
+    public static void executeWithWaitCondition(RowConsumer consumer,
```
Comment (seut) [inline server/src/main/java/io/crate/planner/statement/CopyPlan.java:39]: Same here, please avoid including argument description inside the method name if not really needed.
```suggestion
    public static void execute(RowConsumer consumer,
```

File: server/src/test/java/io/crate/planner/statement/CopyPlannerTest.java [hunk 11]
```diff
@@ -0,0 +1,64 @@
...
+    }
+
+    @Test
+    public void testCopyPlanNoWaitForCompletion() throws Exception {
+        TestingRowConsumer consumer = new TestingRowConsumer();
+        CopyPlan.executeWithWaitCondition(consumer, plannerContext.transactionContext(), (a,b)-> {}, false);
+        assertThat((Long)consumer.getResultWithTimeout(1).get(0)[0], is(-1L));
+    }
+
+    @Test(expected = TimeoutException.class)
```
Comment (seut) [inline server/src/test/java/io/crate/planner/statement/CopyPlannerTest.java:58]: Can we use `assertThrowsMatches` instead?

File: server/src/main/java/io/crate/planner/statement/CopyFromPlan.java (hunk 9 above)
Comment (disideris) [reply server/src/main/java/io/crate/planner/statement/CopyFromPlan.java:356]: @seut can you have another look please?

Yes the try/catch is used only for a possible class cast exception. 
`CopyFromPlannerTest` has tests that check that plan is instance of the right classes and `CopyToPlannerTest` assumes correct classes and uses casts.

So can we remove completely any try/catch/assertion or use assertion as you suggested?

Comment (mfussenegger) [reply server/src/main/java/io/crate/planner/statement/CopyFromPlan.java:356]: I think we should try to get rid of the casts altogether.

You can change the way some methods interact and how the code is structured a bit. For example you could call `bind` first to get access to the `BoundCopyFrom` and then call  `planCopyFromExecution` with that. This would allow you to access the `wait_for_completion` option directly.

File: server/src/main/java/io/crate/planner/statement/CopyFromPlan.java [hunk 12]
```diff
@@ -85,7 +86,7 @@
 import static io.crate.analyze.CopyStatementSettings.settingAsEnum;
 import static io.crate.analyze.GenericPropertiesConverter.genericPropertiesToSettings;
 
-public final class CopyFromPlan implements Plan {
+public final class CopyFromPlan extends CopyPlan implements Plan {
```
Comment (mfussenegger) [inline server/src/main/java/io/crate/planner/statement/CopyFromPlan.java:89]: No need to make `CopyPlan` abstract and use inheritance here to re-use a static method.

File: server/src/test/java/io/crate/integrationtests/CopyIntegrationTest.java (hunk 2 above)
Comment (BaurzhanSakhariev) [reply server/src/test/java/io/crate/integrationtests/CopyIntegrationTest.java]: >never had an issue with manual tests.

if it's always -1 with manual testing it might be wrong test setup (then it's ok to have only unit tests if it's backed by real manual testing). But I'm really wondering what could be wrong in such small test case - I hope it doesn't indicate than it highlights some bug which is somehow slipped away in manual testing.

File: server/src/main/java/io/crate/planner/statement/CopyPlan.java [hunk 13]
```diff
@@ -0,0 +1,51 @@
...
+
+import io.crate.data.CollectingRowConsumer;
+import io.crate.data.InMemoryBatchIterator;
+import io.crate.data.Row;
+import io.crate.data.Row1;
+import io.crate.data.RowConsumer;
+import io.crate.metadata.TransactionContext;
+
+
+public class CopyPlan {
```
Comment (BaurzhanSakhariev) [inline server/src/main/java/io/crate/planner/statement/CopyPlan.java:37]: maybe we can make it abstract, make `CopyFromPlan` and `CopyToPlan` subclasses of this class, make execute protected and call it from children - not a big logical difference just kinda natural to expect that `CopyPlan` is the parent.

Comment (disideris) [reply server/src/main/java/io/crate/planner/statement/CopyPlan.java:37]: @BaurzhanSakhariev I had it this way but see review comment of @mfussenegger above:

> No need to make CopyPlan abstract and use inheritance here to re-use a static method.

Comment (BaurzhanSakhariev) [reply server/src/main/java/io/crate/planner/statement/CopyPlan.java:37]: Didn't see it - well, matter of taste probably - I stumbled over this during review, probably because of the naming - with static approach I would use some different name. No strong opinion here, fine as is.

File: server/src/main/java/io/crate/planner/statement/CopyPlan.java [hunk 14]
```diff
@@ -0,0 +1,51 @@
...
+
+    public static void execute(RowConsumer consumer,
+                               TransactionContext txnCtx,
+                               BiConsumer<RowConsumer, TransactionContext> biConsumer,
+                               boolean waitForCompletion) {
+
+        if (waitForCompletion) {
+            biConsumer.accept(consumer, txnCtx);
+        } else {
+            biConsumer.accept(new CollectingRowConsumer<>(Collectors.mapping(Row::materialize, Collectors.toList())), txnCtx);
```
Comment (BaurzhanSakhariev) [inline server/src/main/java/io/crate/planner/statement/CopyPlan.java:47]: similar to https://github.com/crate/crate/pull/12245#discussion_r830114135 now its same consumers but anyway have concern:

biconsumer here is JobLauncher execute and we are passing RowConsumer.
in case of exception JobLauncer [will call](https://github.com/crate/crate/blob/b8f5bfc1ae1d34b0426f2a01217fde663505225b/server/src/main/java/io/crate/execution/engine/JobLauncher.java#L154-L156) `passedRowConsumer.accept(null, throwable);`
and then we immediately call `consumer.accept(InMemoryBatchIterator.of(Row1.ROW_COUNT_UNKNOWN, SENTINEL), null);`
basically, it's 2 consumer.accept calls in case of exception  - not sure if it's a problem. @mfussenegger  wdyt?

Comment (disideris) [reply server/src/main/java/io/crate/planner/statement/CopyPlan.java:47]: Thanks for mentioning it again. I am not sure. I thought @seut that made the original review would have an idea.

Comment (disideris) [reply server/src/main/java/io/crate/planner/statement/CopyPlan.java:47]: @BaurzhanSakhariev in some tests i made we have the following behavior:
In the case of `wait_for_completion = true`, an exception is returned to the user.
In the case of `wait_for_completion = false`, -1 is returned to the user but the copy operation has failed silently.

Comment (seut) [reply server/src/main/java/io/crate/planner/statement/CopyPlan.java:47]: @disideris @BaurzhanSakhariev 
> In the case of wait_for_completion = true, an exception is returned to the user.
> In the case of wait_for_completion = false, -1 is returned to the user but the copy operation is failing silently.

I think this is the expected behaviour.

File: server/src/main/java/io/crate/planner/statement/CopyPlan.java (hunk 13 above)
Comment (seut) [reply server/src/main/java/io/crate/planner/statement/CopyPlan.java:37]: Please don't create inheritance if it can be avoided (see numerous post about "why inheritance is bad").
Also I wonder if it is really worth to create a dedicated class for just 1 static method shared between to similar classes. Maybe move to `CopyFromPlan` and reuse at `CopyToPlan`?
If you prefer to have this new class, make it `final` please.

File: server/src/test/java/io/crate/planner/statement/CopyPlannerTest.java (hunk 11 above)
Comment (seut) [reply server/src/test/java/io/crate/planner/statement/CopyPlannerTest.java:58]: whats wrong with:

```
assertThrowsMatches(
    () -> execute("select name from users where _version = 1"),
    instanceOf(TimeoutException.class)
);
```
anyway, you can also leave it like that as we're not interested in validating the exception message so `assertThrowsMatches` may not create additional value.

**Comment by disideris** [comment]
retest this please

**Review by BaurzhanSakhariev — APPROVED** [review]
👍 Don't have any more comments - regarding integration test - if it works manually and with unit tests probably it's enough and doesn't worth to dig deep into debugging integration test setup.

File: server/src/main/java/io/crate/planner/statement/CopyPlan.java (hunk 13 above)
Comment (mfussenegger) [reply server/src/main/java/io/crate/planner/statement/CopyPlan.java]: Inheritance can be used if there is an appropriate relationship between the different classes, but COPY FROM and COPY TO are rather separate - that they share a common name doesn't imply a bigger relationship then say COPY TO has with INSERT INTO.

Making a single static function accessible for multiple components is a bad use of inheritance because inheritance  couples components bi-directional.

In this particular case I think putting it in `CopyPlan` doesn't even fit, because the functionality to (not) wait for completion is not tied to COPY FROM or COPY TO either, but it is a general concept. It could even fit into the `JobLauncher`.
In some cases it is fine to put general functionality into concrete components if it isn't used anywhere else to keep the API surface smaller, but in this case we're even adding a new abstraction, so not sure if that point holds.

File: server/src/main/java/io/crate/planner/statement/CopyPlan.java [hunk 15]
```diff
@@ -0,0 +1,51 @@
...
```
Comment (mfussenegger) [inline server/src/main/java/io/crate/planner/statement/CopyPlan.java:47]: Do we have to materialize the results into a list here? In my earlier comment I mentioning using a sum/count collector.

File: server/src/test/java/io/crate/execution/engine/JobLauncherWaitForCompletionTest.java [hunk 16]
```diff
@@ -0,0 +1,64 @@
...
+
+public class JobLauncherWaitForCompletionTest extends CrateDummyClusterServiceUnitTest {
+
+    private PlannerContext plannerContext;
+
+    @Before
+    public void setupExecutor() throws IOException {
+        SQLExecutor e = SQLExecutor.builder(clusterService)
+            .addTable(USER_TABLE_DEFINITION)
+            .addTable("create table t1 (a string, x int, i int)")
```
Comment (seut) [inline server/src/test/java/io/crate/execution/engine/JobLauncherWaitForCompletionTest.java:46]: Afaik we only need the plannerContext from the executor so any table setup is not needed.

File: server/src/main/java/io/crate/planner/statement/CopyToPlan.java [hunk 17]
```diff
@@ -115,24 +124,21 @@ public void executeOrFail(DependencyCarrier executor,
...
             .fromPlan(executionPlan, executor.localNodeId());
-        executor.phasesTaskFactory()
-            .create(plannerContext.jobId(), List.of(nodeOpTree))
-            .execute(consumer, plannerContext.transactionContext());
+
+        JobLauncher jobLauncher = executor.phasesTaskFactory()
+            .create(plannerContext.jobId(), List.of(nodeOpTree));
+
+        JobLauncher.execute(consumer, plannerContext.transactionContext(), jobLauncher::execute,
+                            boundedCopyTo.withClauseOptions().getAsBoolean("wait_for_completion", true));
```
Comment (seut) [inline server/src/main/java/io/crate/planner/statement/CopyToPlan.java:132]: nitpick/minor code style suggestion/preference as this increases readability and produces better diffs on possible future changes. our general pattern here is, either all in one line if fitting into our max char of 120, or each arg in a newline.
```suggestion
        JobLauncher.execute(
            consumer,
            plannerContext.transactionContext(),
            jobLauncher::execute,
            boundedCopyTo.withClauseOptions().getAsBoolean("wait_for_completion", true)
        );
```

File: server/src/main/java/io/crate/planner/statement/CopyFromPlan.java [hunk 18]
```diff
@@ -118,9 +128,11 @@ public void executeOrFail(DependencyCarrier dependencies,
...
             .fromPlan(plan, dependencies.localNodeId());
 
-        dependencies.phasesTaskFactory()
-            .create(plannerContext.jobId(), List.of(nodeOpTree))
-            .execute(consumer, plannerContext.transactionContext());
+        JobLauncher jobLauncher = dependencies.phasesTaskFactory()
+            .create(plannerContext.jobId(), List.of(nodeOpTree));
+
+        JobLauncher.execute(consumer, plannerContext.transactionContext(), jobLauncher::execute,
+                     boundedCopyFrom.settings().getAsBoolean("wait_for_completion", true));
```
Comment (seut) [inline server/src/main/java/io/crate/planner/statement/CopyFromPlan.java:135]: see my other suggestion about code style.

**Review by seut — APPROVED** [review]
Lgtm, added a few last, quite minor suggestions.

File: server/src/main/java/io/crate/execution/engine/JobLauncher.java [hunk 19]
```diff
@@ -142,6 +147,18 @@
...
+                               TransactionContext txnCtx,
+                               BiConsumer<RowConsumer, TransactionContext> biConsumer,
+                               boolean waitForCompletion) {
+        if (waitForCompletion) {
+            biConsumer.accept(consumer, txnCtx);
+        } else {
+            biConsumer.accept(new CollectingRowConsumer<>(Collectors.counting()), txnCtx);
+            consumer.accept(InMemoryBatchIterator.of(Row1.ROW_COUNT_UNKNOWN, SENTINEL), null);
+        }
+    }
```
Comment (mfussenegger) [inline server/src/main/java/io/crate/execution/engine/JobLauncher.java:160]: I think now that it is in the `JobLauncher` it would make sense to make it non-static and remove the `biConsumer` argument because it is currently always parameterised as `jobLauncher::execute`

Comment (disideris) [reply server/src/main/java/io/crate/execution/engine/JobLauncher.java:160]: @mfussenegger Then the unit test could be tricky. I would need to have an instance of JobLauncher !?

Comment (disideris) [reply server/src/main/java/io/crate/execution/engine/JobLauncher.java]: @mfussenegger I made the new `execute` non-static but kept the biconsumer as @seut suggested in early review for easy unit testing.

**Comment by disideris** [comment]
retest this please

File: server/src/main/java/io/crate/execution/engine/JobLauncher.java (hunk 19 above)
Comment (mfussenegger) [reply server/src/main/java/io/crate/execution/engine/JobLauncher.java]: That's kinda taking the disadvantages of both - you still need the `JobLauncher` instance but don't take advantage of having access to the instance variables.

I find the API/abstraction rather confusing in this state. The method is called `execute` but what it does actually depends on generic `biConsumer` parameter.

You could do something like this:

```patch
diff --git a/libs/dex/src/testFixtures/java/io/crate/testing/TestingRowConsumer.java b/libs/dex/src/testFixtures/java/io/crate/testing/TestingRowConsumer.java
index fdfbf954174..46af3c81cb2 100644
--- a/libs/dex/src/testFixtures/java/io/crate/testing/TestingRowConsumer.java
+++ b/libs/dex/src/testFixtures/java/io/crate/testing/TestingRowConsumer.java
@@ -61,13 +61,13 @@ public CompletableFuture<?> completionFuture() {
 
     public List<Object[]> getResult() throws Exception {
         return getResult(10);
     }
 
-    public List<Object[]> getResult(int timeout) throws Exception {
+    public List<Object[]> getResult(int timeout_in_ms) throws Exception {
         try {
-            return consumer.completionFuture().get(timeout, TimeUnit.SECONDS);
+            return consumer.completionFuture().get(timeout_in_ms, TimeUnit.MILLISECONDS);
         } catch (ExecutionException e) {
             Throwable cause = e.getCause();
             if (cause != null) {
                 Exceptions.rethrowUnchecked(cause);
             }
diff --git a/server/src/main/java/io/crate/execution/engine/JobLauncher.java b/server/src/main/java/io/crate/execution/engine/JobLauncher.java
index dc88413ed84..19e83483eb0 100644
--- a/server/src/main/java/io/crate/execution/engine/JobLauncher.java
+++ b/server/src/main/java/io/crate/execution/engine/JobLauncher.java
@@ -99,11 +99,11 @@
  *            N1<----------------+  and passed into a DistResultRXTask
  *             |
  *          BatchConsumer
  * </pre>
  **/
-public final class JobLauncher {
+public class JobLauncher {
 
     private final TransportJobAction transportJobAction;
     private final TransportKillJobsNodeAction transportKillJobsNodeAction;
     private final List<NodeOperationTree> nodeOperationTrees;
     private final UUID jobId;
@@ -145,18 +145,15 @@ public final class JobLauncher {
                 }
             }
         }
     }
 
-    public void execute(RowConsumer consumer,
-                        TransactionContext txnCtx,
-                        BiConsumer<RowConsumer, TransactionContext> biConsumer,
-                        boolean waitForCompletion) {
+    public void execute(RowConsumer consumer, TransactionContext txnCtx, boolean waitForCompletion) {
         if (waitForCompletion) {
-            biConsumer.accept(consumer, txnCtx);
+            execute(consumer, txnCtx);
         } else {
-            biConsumer.accept(new CollectingRowConsumer<>(Collectors.counting()), txnCtx);
+            execute(new CollectingRowConsumer<>(Collectors.counting()), txnCtx);
             consumer.accept(InMemoryBatchIterator.of(Row1.ROW_COUNT_UNKNOWN, SENTINEL), null);
         }
     }
 
     public void execute(RowConsumer consumer, TransactionContext txnCtx) {
diff --git a/server/src/test/java/io/crate/execution/engine/JobLauncherWaitForCompletionTest.java b/server/src/test/java/io/crate/execution/engine/JobLauncherWaitForCompletionTest.java
index d07a32ecb27..06e51d83ca1 100644
--- a/server/src/test/java/io/crate/execution/engine/JobLauncherWaitForCompletionTest.java
+++ b/server/src/test/java/io/crate/execution/engine/JobLauncherWaitForCompletionTest.java
@@ -20,15 +20,27 @@
  */
 
 package io.crate.execution.engine;
 
 import static org.hamcrest.core.Is.is;
+
 import java.io.IOException;
 import java.util.List;
+import java.util.UUID;
 import java.util.concurrent.TimeoutException;
+
+import org.elasticsearch.indices.IndicesService;
 import org.junit.Before;
 import org.junit.Test;
+import org.mockito.Mockito;
+
+import io.crate.data.RowConsumer;
+import io.crate.execution.jobs.JobSetup;
+import io.crate.execution.jobs.TasksService;
+import io.crate.execution.jobs.kill.TransportKillJobsNodeAction;
+import io.crate.execution.jobs.transport.TransportJobAction;
+import io.crate.metadata.TransactionContext;
 import io.crate.planner.PlannerContext;
 import io.crate.test.integration.CrateDummyClusterServiceUnitTest;
 import io.crate.testing.SQLExecutor;
 import io.crate.testing.TestingRowConsumer;
 
@@ -40,32 +52,37 @@ public class JobLauncherWaitForCompletionTest extends CrateDummyClusterServiceUn
     @Before
     public void setupExecutor() throws IOException {
         SQLExecutor e = SQLExecutor.builder(clusterService).build();
         plannerContext = e.getPlannerContext(clusterService.state());
         jobLauncher = new JobLauncher(
-            null,
-            null,
-            null,
-            null,
-            null,
-            null,
-            null,
+            UUID.randomUUID(),
+            clusterService,
+            Mockito.mock(JobSetup.class),
+            Mockito.mock(TasksService.class),
+            Mockito.mock(IndicesService.class),
+            Mockito.mock(TransportJobAction.class),
+            Mockito.mock(TransportKillJobsNodeAction.class),
             List.of(),
             false,
-            null);
+            THREAD_POOL.generic()
+        ) {
+
+            @Override
+            public void execute(RowConsumer consumer, TransactionContext txnCtx) {
+            }
+        };
     }
 
     @Test
-    public void testCopyPlanNoWaitForCompletion() throws Exception {
+    public void test_execute_without_wait_for_completion_yields_immediate_result() throws Exception {
         TestingRowConsumer consumer = new TestingRowConsumer();
-        jobLauncher.execute(consumer, plannerContext.transactionContext(), (a,b)-> {}, false);
-        assertThat((Long)consumer.getResult(1).get(0)[0], is(-1L));
-
+        jobLauncher.execute(consumer, plannerContext.transactionContext(), false);
+        assertThat((Long)consumer.getResult(50).get(0)[0], is(-1L));
     }
 
-    @Test(expected = TimeoutException.class)
-    public void testCopyPlanWaitForCompletion() throws Exception {
+    @Test
+    public void test_execute_with_wait_for_completion_timeouts() throws Exception {
         TestingRowConsumer consumer = new TestingRowConsumer();
-        jobLauncher.execute(consumer, plannerContext.transactionContext(), (a,b)-> {}, true);
-        consumer.getResult(1);
+        jobLauncher.execute(consumer, plannerContext.transactionContext(), true);
+        assertThrows(TimeoutException.class, () -> consumer.getResult(100));
     }
 }
 
 ```
